Implement a py-classindex and py-funindex additionally to the default py-modindex

Naereen opened the pull request:
Subject: Implement a `py-classindex` and `py-funindex` additionally to the default `py-modindex`.

### Feature
- Add two index pages for the Python domain: `classindex` and `funindex`
- Suggestion to fix #3178. Not fully mature yet, but it works already

### Purpose
- Provides a cleaner and fully static solution to [this question](https://stackoverflow.com/questions/12150491/toc-list-with-all-classes-generated-by-automodule-in-sphinx-docs)

### Detail
- The code in `sphinx.domains.python.PythonClassIndex` and `sphinx.domains.python.PythonFunIndex` is very similar  to `sphinx.domains.python.PythonModuleIndex`
- So the look of the resulting index pages are very similar, but they do the job I think (see below for screenshots in the comments)
- I added the new feature in the `CHANGES` file
- It adds two config value, `classindex_common_prefix` and `funindex_common_prefix` that are exactly like `modindex_common_prefix`, and are documented and initialized similarly
- I updated manually the French ``.po`` file, but cannot do the same for other languages

### Other things to do ?
- Add a test ?
- Add the label links to ``classindex`` and ``funindex`` after the label link to ``modindex`` in the default template (for Python domain only) ?
- Do the same for ``data`` (constants) ?

### Relates
- #3178

Naereen committed:
Suggestion to fix #3178. Not fully mature yet, but it works already.

Naereen commented:
Here are some screenshots:

- The added `classindex` and `funindex` label link on a `.rst` page: (not translated yet)
![demo main index](https://cloud.githubusercontent.com/assets/11994719/21464500/32957dc0-c97f-11e6-87bc-af2049f56bec.png)

- Example of a class index with some classes, order by modules (exactly as the default mod index):
![demo class index](https://cloud.githubusercontent.com/assets/11994719/21464507/627143a8-c97f-11e6-9fd3-41a5058f81de.png)

- 2 examples of a fun index with some functions, order by modules (exactly as the default mod index):
![demo fun index](https://cloud.githubusercontent.com/assets/11994719/21464512/7b564fb2-c97f-11e6-80e4-75add3b23e35.png)
![demo fun index 2](https://cloud.githubusercontent.com/assets/11994719/21464523/db5c9f1a-c97f-11e6-833a-fe5735b8f711.png)

Naereen committed:
Probably better to display a '()' at the end of the function name, like in the genindex page.

Naereen commented:
Note: I don't think the travis build failed because of my modification, based on what I saw.

Naereen commented:
Here is a demo of how it looks on a "real" (tiny) project:

- fun index: [example 1](http://perso.crans.org/besson/infoMP/TPs/solutions/py-funcindex.html), [example 2](http://perso.crans.org/besson/infoMP/DSs/solutions/py-funcindex.html),
- class index: [example 3](http://perso.crans.org/besson/infoMP/TDs/solutions/py-classindex.html),
- more indexes, for all possible Python object types: [example 4](http://perso.crans.org/besson/infoMP/TPs/solutions/#table-des-matieres) (in French)

Naereen commented:
@tk0miya what do you think?

By the way, merry Xmas to all ! :christmas_tree: :santa: :gift: !

tk0miya commented on doc/config.rst:
```diff
@@ -450,6 +450,26 @@ Project information
+
+   .. versionadded:: 1.6
+
+.. confval:: funindex_common_prefix
```
Sphinx uses "func" as abbreviation of the function like `:py:func:`.
So I prefer to rename this to `funcindex_common_prefix`.

Naereen replied:
OK, I will fix that.

tk0miya commented on sphinx/domains/python.py:
```diff
@@ -641,6 +641,152 @@ def generate(self, docnames=None):
         return sorted_content, collapse
 
 
+class PythonClassIndex(Index):
```
It seems these three index classes are very similar.
How about making a common parent class?
It would be nice to keep code simple :-)

Naereen replied:
I would rather write a function generating a class (a functor), to be able to easily have subclasses of `Index` for classes, functions, data, attributes, methods, classmethods, staticmethods etc.

Naereen replied:
Cf. b9d91d3
It seems cleaner.

tk0miya commented on sphinx/locale/fr/LC_MESSAGES/sphinx.po:
```diff
@@ -1,7 +1,7 @@
-# Translations template for Sphinx.
+# French translations for Sphinx.
```
Hmm? Did you translate this manually?
We use transifex.com to manage translations.
So you don't change *.po file manually (at least, in master branch).

Naereen replied:
OK, I knew it was a bad idea...

Naereen replied:
I reverted the changes, in c6cbd92. Sorry.

tk0miya reviewed (changes requested):
As a quick look, it looks good to me.
BTW, could you add testcase?

tk0miya commented on sphinx/domains/python.py:
```diff
@@ -641,6 +641,152 @@ def generate(self, docnames=None):
+        ignores = sorted(ignores, key=len, reverse=True)
+        # list of all classes, sorted by class name
+        unsorted_classes = [(refname, (docname, type))
+                            for (refname, (docname, type)) in iteritems(self.domain.data['objects'])
```
In addition, Mr flake8 warns this line is too long. Please fold the line :-)
Line 735 is also.
https://travis-ci.org/sphinx-doc/sphinx/jobs/186467941#L441

Naereen replied:
OK, thanks. I fixed that.

tk0miya commented:
@shimizukawa Could you give us advice?
I'd reviewed this in coding level, but I could not determine this is really needed or not to sphinx-core.

Naereen committed:
Change 'funindex( to 'funcindex' (cf https://github.com/sphinx-doc/sphinx/pull/3278#discussion_r93812085)

Naereen committed:
Add a functor makePythonCustomIndex(...) -> subclass of Index, to make all the different indeces easily.

Naereen committed:
Remove DEBUG prints, and fold lines to conform flake8 style.

Naereen commented:
Maybe it's too much, but I added the possibility to generate (and link to, with a label link) indeces of all these Python object types:

- `class`
- `function`
- `data`
- `exception`
- `method`
- `classmethod`
- `staticmethod`
- `attribute`

Example: [here (in French)](http://perso.crans.org/besson/infoMP/TPs/solutions/#table-des-matieres).

Naereen committed:
Sorry, typo in ./sphinx/domains/python.py:646:31: E127 continuation line over-indented for visual indent

Naereen committed:
Revert changes on sphinx/locale/fr/LC_MESSAGES/sphinx.po back to 673418e.

- I will update the French translation on Transifex, afterward.

Naereen reviewed (commented):
I changed the way to customize this, and `'()'` is now added to the displayed name of functions, classmethods, staticmethods, and methods in their respective index.

Naereen committed:
Added default values and documentation for the config values

- `dataindex_common_prefix`
- `exceptionindex_common_prefix`
- `methodindex_common_prefix`
- `classmethodindex_common_prefix`
- `staticmethodindex_common_prefix`
- `attributeindex_common_prefix`

Naereen commented:
I am not sure how to add test cases for these changes. Any hint please?

shimizukawa commented:
Thank you for your PR. I also think that it is useful when implementing a large library to have a list of classes and functions as individual index pages. BTW, Sphinx is a tool suitable for document creation for Python, however it is also open to applications other than Python. Therefore, I would like to avoid adding Python specific settings and mechanisms to the global space drastically.

* +1: I think categorized index page is useful
* -1: Adding config and label specific to python domain in global space (it is not good that the existing `py-modindex` is specific to the Python domain but implemented in std)
* -1: There is no link to the new index pages and there is no documentation, so it is difficult for users to use it.
* Suggestion: You should have a representative index page and the page have links to each index pages
* Proposal: At the beginning, let's make it as an independent Sphinx extension. It can be grown in shorter cycles than Sphinx and you can act quickly based on user feedback.

Naereen committed:
Few modifications:

- Changed names of config values to be Python specific: py_classindex_common_prefix etc
- Moved documentation of these config values to a new section in doc/config.rst: « Options for the Python domain »
- Shorter names of labels and anonlabels, conform to [the documentation](http://www.sphinx-doc.org/en/1.5.1/domains.html#cross-referencing-python-objects)
- Use the config value [add_function_parentheses](http://www.sphinx-doc.org/en/1.5.1/config.html#confval-add_function_parentheses) to know if changeFuncNames(self, s) should be s or s + '()'

Naereen committed:
Flake8 is not my friend. Yet.

Naereen committed:
Proposal for links to (Python specific) additional index pages (classindex, funcindex etc) in the bottom of index pages for the basic theme (demo).

Naereen committed:
Typo.

Naereen commented:
Thanks @shimizukawa for your insights.
I know and understand your wish to keep the Python-specific things away from the generic Sphinx tool.

- About first -1 : I tried to define the `labels` and `anonlabels` for `py-classindex` (& all others) in the `python.py` domain file rather than the `std.py` domain file, but it ain't no good as the labels are mainly intended to be used in `.rst` files (std domain) rather than in Python docstrings...
- About second -1 and the suggestion, dff3d4b is a suggestion of how to had links to these index pages. Here is an example:
![demo all index](https://cloud.githubusercontent.com/assets/11994719/21484838/9f2701ae-cb99-11e6-9774-87d214060a4e.png)
- By the way, I found a surprising feature, on the `basic` theme, all the new index pages are added to the `relbar2` block. It might get too much if all the index pages are used, maybe `themes/basic/layout.html` should be modified to only include some links and not all (`{%- for rellink in rellinks %}`, line 31).

About your proposal, I agree it could be more efficient, but I prefer to be clear right away: I don't and won't have time to neither develop the extension nor maintain it. I know it takes some time, just to have either a PyPI package, or a merge in sphinxcontrib.
Sorry.

Naereen committed:
More consistent constant name.

Naereen commented:
Any update ?

shimizukawa commented:
We also do not have much time. I think that the feature you want is wonderful. However, I feel that the current implementation you write complicates Sphinx's code base. For this reason, once we adopt the current proposal code, it is difficult for us to maintain in the future.

tk0miya commented:
@shimizukawa Could you close this? As a quick look, we decided not to merge this. So this should be closed.
